### ISSUE TITLE
Syntax Error - unexpected identifier - with api_keys in truffle config (truffle plugin verify)

### ISSUE THREAD

**Issue by villarode** [rkalis/truffle-plugin-verify#127]
Hello guys I am running into a lil prob using automatic contract verification plugin:

I have everyhing set up in my truffle config:

plugins: [
'truffle-plugin-verify'
]
api_keys: {
etherscan: process.env.ETHERSCAN_API_KEY
}
};

However when I run ::

$ yarn truffle deploy --network rinkeby
C:\Users\User\OS721_Rinkeby\truffle-config.js:120
api_keys: {
^^^^^^^^

SyntaxError: Unexpected identifier

Any idea why I am getting this Syntax Error? Cheers in advance!

**Comment by rkalis** [comment]
From looking at this it appears you're missing a comma after the `]`.

**Comment by villarode** [comment]
Yes mate, you are absolutely tight, cheers for the prompt answer!)

**Comment by villarode** [comment]
Hello again, I have another question if I may) also related to verifying contracts using truffle verify plugin.

The name of my deploy compilation target (sitting in build/contracts... ) is Villa_Rode_AL,, this is the name of my contract for verify plugin, right?

So this is what I'm getting with and without the contract digits after the ContractName:

User@WIN-7DVVFALHNLB MINGW64 ~/OS721_Rinkeby (master)
$ yarn truffle run verify Villa_Rode_AL --network rinkeby
Verifying Villa_Rode_AL
No instance of contract Villa_Rode_AL found for network id 4
Failed to verify 1 contract(s): Villa_Rode_AL

User@WIN-7DVVFALHNLB MINGW64 ~/OS721_Rinkeby (master)
$ yarn truffle run verify Villa_Rode_AL@0xb5d9837aE53020706e0E7a0b5B586499d1f3a5aD --network rinkeby
Verifying Villa_Rode_AL@0xb5d9837aE53020706e0E7a0b5B586499d1f3a5aD
Fail - Unable to verify
Failed to verify 1 contract(s): Villa_Rode_AL@0xb5d9837aE53020706e0E7a0b5B586499d1f3a5aD

What am I doing wrong?) Cheers in advance!)

**Comment by rkalis** [comment]
Did you deploy the contract using Truffle as well (`truffle migrate ...`)? Was the contract deployed by another smart contract?

**Comment by villarode** [comment]
Yes, it was deployed using truffle as well - as a matter of fact I re-deployed it before running verify:

$ yarn truffle deploy --network rinkeby

Compiling your contracts...
===========================
> Compiling @openzeppelin\contracts\access\Ownable.sol
> Compiling @openzeppelin\contracts\token\ERC721\ERC721.sol
> Compiling @openzeppelin\contracts\token\ERC721\IERC721.sol
> Compiling @openzeppelin\contracts\token\ERC721\IERC721Receiver.sol
> Compiling @openzeppelin\contracts\token\ERC721\extensions\ERC721Enumerable.sol
> Compiling @openzeppelin\contracts\token\ERC721\extensions\IERC721Enumerable.sol
> Compiling @openzeppelin\contracts\token\ERC721\extensions\IERC721Metadata.sol
> Compiling @openzeppelin\contracts\utils\Address.sol
> Compiling @openzeppelin\contracts\utils\Context.sol
> Compiling @openzeppelin\contracts\utils\Strings.sol
> Compiling @openzeppelin\contracts\utils\introspection\ERC165.sol
> Compiling @openzeppelin\contracts\utils\introspection\IERC165.sol
> Compiling @openzeppelin\contracts\utils\math\SafeMath.sol
> Compiling .\contracts\ApprovedSpenderContract.sol
> Compiling .\contracts\ERC721Tradable.sol
> Compiling .\contracts\Migrations.sol
> Compiling .\contracts\Villa_Rode_AL_OS721_Rinkeby.sol
> Compiling .\contracts\common\meta-transactions\ContentMixin.sol
> Compiling .\contracts\common\meta-transactions\EIP712Base.sol
> Compiling .\contracts\common\meta-transactions\Initializable.sol
> Compiling .\contracts\common\meta-transactions\NativeMetaTransaction.sol
> Artifacts written to C:\Users\User\OS721_Rinkeby\build\contracts
> Compiled successfully using:
   - solc: 0.8.10+commit.fc410830.Emscripten.clang

Network up to date.
Network up to date.

**Comment by villarode** [comment]
The contract was not deployed thru another contract, I basically did all the deployment from scratch installing nvm, yarn etc.... Pretty much in line with OpenSea documentation for beginners (Or Alchemy documentation, for that purpose)..

**Comment by villarode** [comment]
That was by the way my original .sol file that I used to create the contract:

// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

import "./ERC721Tradable.sol";
import "@openzeppelin/contracts/access/Ownable.sol";

/**
 * @title Villa_Rode_AL_OS721_Rinkeby
 * Creature - a contract for my non-fungible creatures.
 */
contract Villa_Rode_AL is ERC721Tradable {
    constructor(address _proxyRegistryAddress)
        ERC721Tradable("Villa_Rode_AL_OS721_Rinkeby", "OSC", _proxyRegistryAddress)
    {}

    function baseTokenURI() override public pure returns (string memory) {
        return "https://gateway.pinata.cloud/ipfs/QmQ4btGqRKfLwfJgAuh7k3cAqRDQZj1WTV1L8f9EhdunHQ/";
    }

    function contractURI() public pure returns (string memory) {
        return "https://creatures-api.opensea.io/contract/opensea-creatures";
  }
}

**Comment by villarode** [comment]
Funny thing actually is that

https://testnets.opensea.io/assets/0xb5d9837aE53020706e0E7a0b5B586499d1f3a5aD

0xb5d..... being my deployed contract returns error 404 your asset is not found..... It probably has much to do with me not being able to mint any tokens into my contract using Pinata....(

**Comment by rkalis** [comment]
Can you run `truffle migrate --network rinkeby --reset`. That should force a redeployment.

**Comment by villarode** [comment]
I ran a --reset, obviously I got a new contract number which I also whitelisted in Alchemy.

Nothing changed then:

User@WIN-7DVVFALHNLB MINGW64 ~/OS721_Rinkeby (master)
$ yarn truffle run verify Villa_Rode_AL --network rinkeby
Verifying Villa_Rode_AL
No instance of contract Villa_Rode_AL found for network id 4
Failed to verify 1 contract(s): Villa_Rode_AL

User@WIN-7DVVFALHNLB MINGW64 ~/OS721_Rinkeby (master)
$ yarn truffle run verify Villa_Rode_AL@0x95a0f00c4ca7c3d107455b9215621b17c24f217f --network rinkeby
Verifying Villa_Rode_AL@0x95a0f00c4ca7c3d107455b9215621b17c24f217f
Fail - Unable to verify
Failed to verify 1 contract(s): Villa_Rode_AL@0x95a0f00c4ca7c3d107455b9215621b17c24f217f

**Comment by villarode** [comment]
What especially spuns me is that no instance of the contract is found for rinkeby network.... This contract name (Villa_Rode_AL) was specified from starters, in the initial .sol file and obviiusly is created in ...build/contracts/.... on compilation.

**Comment by rkalis** [comment]
Hmm yeah. When you look in the artifact JSON file for that contract and search for `"networks"`, does it show deployed contracts?

e.g.

```
  "networks": {
    "4": {
      "events": {},
      "links": {},
      "address": "0x2C7C07AdB57f3959f28563f9dC79d8bA1736CFb0",
      "transactionHash": "0x1b236389aade5e8e6ece3d7d711b4209c19afaef3423dbe41e583f7314f5c1ce"
    }
  },
  ```

**Comment by villarode** [comment]
amazingly.... in the target file Villa_Rode_AL.json in ....build/contracts/... no, I'm not seeing that.... What I'm seeing is:

"compiler": {
    "name": "solc",
    "version": "0.8.10+commit.fc410830.Emscripten.clang"
  },
  "networks": {},
  "schemaVersion": "3.4.3",
  "updatedAt": "2021-11-17T14:05:41.906Z", etc etc etc

**Comment by villarode** [comment]
even though when the conract was being created I saw all that info:

1_initial_migration.js
======================

   Replacing 'Migrations'
   ----------------------
   > block number:        9658345
   > block timestamp:     1637157965
   > account:             0xD561Fd6763e8cc8c7603C0D7971Bc003DD2Dd17d
   > balance:             21.748416926996584914
   > gas used:            317219 (0x4d723)
   > gas price:           2 gwei
   > value sent:          0 ETH
   > total cost:          0.000634438 ETH

   -------------------------------------
   > Total cost:         0.000634438 ETH


Summary
=======
> Total deployments:   1
> Final cost:          0.000634438 ETH





Starting migrations...
======================
> Network name:    'rinkeby'
> Network id:      4
> Block gas limit: 29970677 (0x1c950f5)


1_initial_migration.js
======================

   Replacing 'Migrations'
   ----------------------
   > transaction hash:    0xe7d996468f8eb6419f9f0d4f3b094faa6d883027970fd2cd008098e1b4eebac4
- Blocks: 0            Seconds: 0
   > Blocks: 0            Seconds: 8
   > contract address:    0x95a0f00C4ca7C3d107455B9215621B17c24F217F
   > block number:        9658346
   > block timestamp:     1637157985
   > account:             0xD561Fd6763e8cc8c7603C0D7971Bc003DD2Dd17d
   > balance:             21.748217567492916205
   > gas used:            333519 (0x516cf)
   > gas price:           2.500000011 gwei
   > value sent:          0 ETH
   > total cost:          0.000833797503668709 ETH


- Saving migration to chain.
   > Saving migration to chain.
   > Saving artifacts
   -------------------------------------
   > Total cost:     0.000833797503668709 ETH

**Comment by villarode** [comment]
Truffle-config seems to be alright::

networks: {
    // Useful for testing. The `development` name is special - truffle uses it by default
    // if it's defined here and no other network is specified at the command line.
    // You should run a client (like ganache-cli, geth or parity) in a separate terminal
    // tab if you use this network and you must also set the `host`, `port` and `network_id`
    // options below to some value.
    //
    // development: {
    //  host: "127.0.0.1",     // Localhost (default: none)
    //  port: 8545,            // Standard Ethereum port (default: none)
    //  network_id: "*",       // Any network (default: none)
    // },
    // Another network with more advanced options...
    // advanced: {
    // port: 8777,             // Custom port
    // network_id: 1342,       // Custom network
    // gas: 8500000,           // Gas sent with each transaction (default: ~6700000)
    // gasPrice: 20000000000,  // 20 gwei (in wei) (default: 100 gwei)
    // from: <address>,        // Account to send txs from (default: accounts[0])
    // websocket: true        // Enable EventEmitter interface for web3 (default: false)
    // },
    // Useful for deploying to a public network.
    // NB: It's important to wrap the provider as a function.
    rinkeby: {
    provider: () => new HDWalletProvider(mnemonic, `https://eth-rinkeby.alchemyapi.io/v2/tnUAyO3Fdms1of94K3TOz6AUCcWVrQGW`),
    network_id: 4,       // Rinkeby's id
    gas: 5500000,        // Ropsten has a lower block limit than mainnet
    // confirmations: 2,    // # of confs to wait between deployments. (default: 0)
    // timeoutBlocks: 200,  // # of blocks before a deployment times out  (minimum/default: 50)
    // skipDryRun: true     // Skip dry run before migrations? (default: false for public nets )
    // },
    // Useful for private networks
    // private: {
    // provider: () => new HDWalletProvider(mnemonic, `https://network.io`),
    // network_id: 2111,   // This network is yours, in the cloud.
    // production: true    // Treats this network as if it was a public net. (default: false)
    }
  },

**Comment by villarode** [comment]
Actually, found it:

"networks": {
    "4": {
      "events": {},
      "links": {},
      "address": "0x95a0f00C4ca7C3d107455B9215621B17c24F217F",
      "transactionHash": "0xe7d996468f8eb6419f9f0d4f3b094faa6d883027970fd2cd008098e1b4eebac4"
    }
  },

It sits in Migrations.json file which is also in .... build/contracts..... directory along with Villa_Rode_AL.json file

**Comment by villarode** [comment]
and now when I try to verify Migrations.json instead of Villa_Rode_AL.json, I'm getting this:

User@WIN-7DVVFALHNLB MINGW64 ~/OS721_Rinkeby (master)
$ yarn truffle run verify Migrations --network rinkeby
Verifying Migrations
Pass - Verified: https://rinkeby.etherscan.io/address/0x95a0f00C4ca7C3d107455B9215621B17c24F217F#code
Successfully verified 1 contract(s).

User@WIN-7DVVFALHNLB MINGW64 ~/OS721_Rinkeby (master)
$ yarn truffle run verify Migrations@0x95a0f00c4ca7c3d107455b9215621b17c24f217f - network rinkeby
Etherscan has no support for network undefined with chain id null

**Comment by villarode** [comment]
Point is the reset contract is successfully verified:

I'm pretty sure by the way it would have worked the same way before the reset if I was doing verify on Migrations.json instaed of on Villa_Rode_AL.json (which I thought was the right contract name to verify).....

By the way why is that we need to do it on Migrations and not on the ContactName itself?)

**Comment by villarode** [comment]
Many thanks for your help!) by the way what's gonna happen to my previous deployment which I reset? Is it onna be still alive somehow? Or can I forget about it altogether?)

**Comment by rkalis** [comment]
Wait it looks like you don't have a deployment script for the `Villa_Rode_AL` contract. By default truffle only has the `1_initial_migration.js` file listed under `migrations/` which deploys `Migrations.sol` if you want to deploy the other contract you need to add an additional deployment script (e.g. `2_deploy_villa_rode_al.js`) with the corresponding deployment code.

**Comment by villarode** [comment]
May be this is all very related to another problem that I'm running into(

I'm trying to create some nfts under the roof of my contract. I have created an nft-metadata.json file with a reference to a picture on Pinata:

{
    "attributes" : [ {
      "trait_type" : "Breed",
      "value" : "Maltipoo"
    }, {
      "trait_type" : "Eye color",
      "value" : "Mocha"
    } ],
    "description" : "The world's most adorable and sensitive pup.",
    "image" : "https://gateway.pinata.cloud/ipfs/QmPaV6zTHbqLPMcZRVLU1hfcB6h9CM2C7U7fz3fgqUckr2",
    "name" : "Ramses"
}

and obviously I uploaded that nft-metadata file to pinata as well. I'm referring to that upload when I compile my contract:

// SPDX-License-Identifier: MIT

pragma solidity ^0.8.0;

import "./ERC721Tradable.sol";
import "@openzeppelin/contracts/access/Ownable.sol";

/**
 * @title Villa_Rode_AL_OS721_Rinkeby
 * Creature - a contract for my non-fungible creatures.
 */
contract Villa_Rode_AL is ERC721Tradable {
    constructor(address _proxyRegistryAddress)
        ERC721Tradable("Villa_Rode_AL_OS721_Rinkeby", "OSC", _proxyRegistryAddress)
    {}

    function baseTokenURI() override public pure returns (string memory) {
        return "https://gateway.pinata.cloud/ipfs/QmQ4btGqRKfLwfJgAuh7k3cAqRDQZj1WTV1L8f9EhdunHQ/";
    }

    function contractURI() public pure returns (string memory) {
        return "https://creatures-api.opensea.io/contract/opensea-creatures";
  }
}

However, I am not seeing that metadata at all. May be it's got much to do what we found out with verification of Migrations instead of Villa_Rode_AL file. My understang is that IF evrything was right I'd be seeing TokenURI in position 13 under Read Contract Information button. However, all I am seeing now is just 2 positions - last_completed_migration (uint256) and owner (with my wallet id).....

What might be the problem? Do I need to migrate to 2_deploy_villa_rode_al.js as you suggested to start seeing the tokens?)

Much appreciated and kind regards

**Comment by rkalis** [comment]
The contract you are looking at now is the "Migrations" contract that is included with Truffle by default. If you add the deploy script for your own contract and then verify it with the plugin you should be able to interact with the correct contract on Etherscan.

**Comment by villarode** [comment]
Cheers mate, much appreciated, I reckon much of cobweb is cleared in my head by now))

**Comment by villarode** [comment]
Is my thinking correct that in migrations file:

const Migrations = artifacts.require("Migrations");

module.exports = function (deployer) {
  deployer.deploy(Migrations);
};

in line artifacts.require("Migrations") Migrations will have to be replaced with my ContractName, e.g. Villa_Rode_AL?

**Comment by rkalis** [comment]
Yeah that's right.

**Comment by rkalis** [comment]
I'm closing this issue. Feel free to open another issue if you run into any verification issues after setting up the deploy scripts.

**Comment by villarode** [comment]
yes of course mate thank you
